### PR TITLE
Introduce ProblemDescription

### PR DESCRIPTION
**Pull request by pranavkm**

### CODE REVIEW

File: samples/MvcSandbox/Controllers/HomeController.cs [hunk 1]
```diff
@@ -14,5 +16,59 @@ public IActionResult Index()
...
+        }
+
+        private const int AvailableFunds = 100;
+
+        [HttpPost("/problem")]
+        public ActionResult<Person> Problem(int cooks)
+        {
+            if (cooks > 10)
+            {
+                return new ProblemDescriptor
```
Comment (rynowak) [inline samples/MvcSandbox/Controllers/HomeController.cs]: Problem**Description**

File: src/Microsoft.AspNetCore.Mvc.Formatters.Json/Internal/ProblemDescriptorConverter.cs [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+{
+    public class ProblemDescriptorConverter : JsonConverter
+    {
+        public override bool CanRead => false;
+
+        public override bool CanConvert(Type objectType) => typeof(ProblemDescriptor).IsAssignableFrom(objectType);
+
+        public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
+        {
+            throw new NotSupportedException();
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Formatters.Json/Internal/ProblemDescriptorConverter.cs]: Need this guy

File: src/Microsoft.AspNetCore.Mvc.Formatters.Json/Internal/ProblemDescriptorConverter.cs [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+                {
+                    var property = properties[i];
+                    if (property.Property.DeclaringType != typeof(ProblemDescriptor))
+                    {
+                        WriteProperty(property.Name, property.GetValue(problem));
+                    }
+                }
+            }
+
+            foreach (var extendedProperty in problem.AdditionalProperties)
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Formatters.Json/Internal/ProblemDescriptorConverter.cs]: Do we need this?

File: src/Microsoft.AspNetCore.Mvc.Core/ProblemDescriptor.cs [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+        /// <summary>
+        /// A URI reference that identifies the specific occurrence of the problem.It may or may not yield further information if dereferenced.
+        /// </summary>
+        public string Instance { get; set; }
+
+        /// <summary>
+        /// Gets or sets additional properties associated with this instance of <see cref="ProblemDescriptor"/>.
+        /// </summary>
+        public IDictionary<string, object> AdditionalProperties { get; } = new Dictionary<string, object>(StringComparer.Ordinal);
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/ProblemDescriptor.cs]: Do we need this?

File: src/Microsoft.AspNetCore.Mvc.Formatters.Xml/Internal/ProblemDescriptiorWrapper.cs [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+namespace Microsoft.AspNetCore.Mvc.Formatters.Xml.Internal
+{
+    /// <summary>
+    /// Wrapper class for <see cref="Mvc.ProblemDescriptor"/> to enable it to be serialized by the xml formatters.
+    /// </summary>
+    [XmlRoot("Error")]
+    public sealed class ProblemDescriptorWrapper : IXmlSerializable, IUnwrappable
+    {
+        // Note: XmlSerializer requires to have default constructor
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Formatters.Xml/Internal/ProblemDescriptiorWrapper.cs]: `requires to have`  👎

**Comment by pranavkm** [comment]
🆙 📅

File: src/Microsoft.AspNetCore.Mvc.Core/ValidationProblemDescription.cs [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+            string GetErrorMessage(ModelError error)
+            {
+                return string.IsNullOrEmpty(error.ErrorMessage) ?
+                    Resources.SerializableError_DefaultError :
+                    error.ErrorMessage;
+            }
+        }
+
+
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Core/ValidationProblemDescription.cs]: Will remove

File: src/Microsoft.AspNetCore.Mvc.Core/ActionResultOfT.cs [hunk 7]
```diff
@@ -50,6 +50,18 @@ public ActionResult(ActionResult result)
             return new ActionResult<TValue>(result);
         }
 
+        public static implicit operator ActionResult<TValue>(ProblemDescription problemDescriptor)
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/ActionResultOfT.cs]: This shouldn't be here, this isn't the right response type for all problem descriptions

File: src/Microsoft.AspNetCore.Mvc.Core/ValidationProblemDescription.cs [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+        /// <summary>
+        /// Gets or sets the validation errors associated with this instance of <see cref="ValidationProblemDescription"/>.
+        /// </summary>
+        public IDictionary<string, object> Errors { get; set; } = new Dictionary<string, object>(StringComparer.Ordinal);
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/ValidationProblemDescription.cs]: this should be `IDictionary<string, string[]>`

File: src/Microsoft.AspNetCore.Mvc.Formatters.Json/Microsoft.AspNetCore.Mvc.Formatters.Json.csproj [hunk 9]
```diff
@@ -13,5 +13,6 @@
 
     <PackageReference Include="Microsoft.AspNetCore.JsonPatch" Version="$(AspNetCoreVersion)" />
     <PackageReference Include="Microsoft.Extensions.ClosedGenericMatcher.Sources" Version="$(AspNetCoreVersion)" PrivateAssets="All" />
+    <PackageReference Include="Microsoft.Extensions.PropertyHelper.Sources" Version="$(AspNetCoreVersion)" PrivateAssets="All" />
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Formatters.Json/Microsoft.AspNetCore.Mvc.Formatters.Json.csproj]: Still needed?

File: src/Microsoft.AspNetCore.Mvc.Core/ValidationProblemDescription.cs (hunk 8 above)
Comment (dougbu) [reply src/Microsoft.AspNetCore.Mvc.Core/ValidationProblemDescription.cs]: Why does this have a setter?

Comment (pranavkm) [reply src/Microsoft.AspNetCore.Mvc.Core/ValidationProblemDescription.cs]: Sure, this could be removed.

Comment (pranavkm) [reply src/Microsoft.AspNetCore.Mvc.Core/ValidationProblemDescription.cs]: Addressed prior to merging this in https://github.com/aspnet/Mvc/commit/a7cc243942c4bdcf8028ad76571432258b1b2533.

**Pull request closed without merging**